scipy/special/_basic.py

```diff
@@ -2065,14 +2072,27 @@ def factorial(n, exact=False):
                     current = un[i + 1]
                     val *= _range_prod(prev, current)
                     out[n == current] = val
+
+            if np.isnan(n).any():
+                out = out.astype(np.float64)
+                out[np.isnan(n)] = np.nan
             return out
     else:
         if np.ndim(n) == 0:
-            return 0 if n < 0 else gamma(n + 1)
+            if not np.isnan(n):
```

Comment:
The check shouldn't be needed here-`gamma` will propagate the NaN correctly.

Reply:
yes, I got overzealous in my NaN handling.